arangod/Pregel/Algos/AIR/GraphFormat.cpp

```diff
@@ -0,0 +1,209 @@
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Heiko Kernbach
+/// @author Lars Maier
+/// @author Markus Pfeiffer
+////////////////////////////////////////////////////////////////////////////////
+
+#include "GraphFormat.h"
+
+#include "Pregel/Algos/AIR/Greenspun/Extractor.h"
+#include "Pregel/Algos/AIR/Greenspun/Interpreter.h"
+
+#include <velocypack/Collection.h>
+#include <velocypack/velocypack-aliases.h>
+
+namespace arangodb {
+namespace pregel {
+namespace algos {
+namespace accumulators {
+
+// Graph Format
+GraphFormat::GraphFormat(application_features::ApplicationServer& server,
+                         std::string const& resultField,
+                         AccumulatorsDeclaration const& globalAccumulatorDeclarations,
+                         AccumulatorsDeclaration const& vertexAccumulatorDeclarations,
+                         CustomAccumulatorDefinitions customDefinitions,
+                         DataAccessDefinition const& dataAccess)
+    : graph_format(server),
+      _resultField(resultField),
+      _globalAccumulatorDeclarations(globalAccumulatorDeclarations),
+      _vertexAccumulatorDeclarations(vertexAccumulatorDeclarations),
+      _customDefinitions(std::move(customDefinitions)),
+      _dataAccess(dataAccess) {}
+
+size_t GraphFormat::estimatedVertexSize() const { return sizeof(vertex_type); }
+size_t GraphFormat::estimatedEdgeSize() const { return sizeof(edge_type); }
+
+void filterDocumentData(VPackBuilder& finalBuilder, VPackSlice const& arraySlice,
+                        arangodb::velocypack::Slice& document) {
+  for (auto&& key : VPackArrayIterator(arraySlice)) {
+    if (key.isString()) {
+      VPackBuilder tmp;
+      VPackBuilder innerBuilder;
+      innerBuilder.openObject();
+      innerBuilder.add(key.copyString(), document.get(key.stringRef()));
+      innerBuilder.close();
+
+      if (!finalBuilder.slice().isObject()) {
+        finalBuilder.openObject();
+        finalBuilder.close();
+      }
+      VPackCollection::merge(tmp, finalBuilder.slice(), innerBuilder.slice(), true, false);
+      finalBuilder.clear();
+      finalBuilder.add(tmp.slice());
+    } else if (key.isArray()) {
+      std::vector<VPackStringRef> path;
+      size_t pathLength = key.length();
+      size_t iterationStep = 0;
+
+      VPackBuilder innerArrayBuilder;
+      innerArrayBuilder.openObject();  // open outer object
```

Comment:
This looks relatively expensive. Wondering if there is some way to speed this up.
For example, if it is guaranteed that `finalBuilder` is always empty when this is called, this method could be sped up a lot.

Reply:
There was a ticket created for that.